### PR TITLE
wire: add an example and document how to use wire with mocks

### PR DESCRIPTION
**Pull request by vangent**

### CODE REVIEW

**Comment by vangent** [comment]
Fixes #477

**Comment by vangent** [comment]
Please review the thread in #477 and see if you can think of a better wait to do this.

I tried to use my last suggestion, wherein initMockApp created the mocked timer itself, and returned it in a wrapper struct, but I don't think that wire supports using the mockTimer as both a "timer" interface *and* the concrete *mockTimer.

**Comment by vangent** [comment]
I'm going to add an example of how to use wire.Bind.

**Comment by zombiezen** [comment]
@vangent Is this ready for review? The bot seems to keep pushing it over to me. 😃

**Comment by vangent** [comment]
Yes, it's ready for review.

**Comment by zombiezen** [comment]
Can you do the reformat in a separate PR? It's really hard for me to see the actual change here.

(It would be so nice to have Contribute Bot do a Markdown linting pass...)

**Comment by vangent** [comment]
> Can you do the reformat in a separate PR? 
Sure, sent you PR #495 .

**Comment by vangent** [comment]
OK, merged, so now only the new section is in the diff.

File: wire/README.md [hunk 1]
```diff
@@ -453,3 +453,24 @@ You may not:
...
+argument types must be the interface types the mocks are mocking. `wire.Build`
+can't include providers for the mocked dependencies without creating conflicts,
+so if you're using provider set(s) you will need to define one that doesn't
+include the mocked types.
+
+#### Approach B: Return the mocks from the injector
+
+Create a new struct that includes the app plus all of the dependencies you want
+to mock. Create a test-only injector that returns this struct, give it providers
+for the concrete mock types, and use `wire.Bind` to tell `wire` that the
```
Comment (zombiezen) [inline wire/README.md]: nit: s/\`wire\`/Wire/

File: wire/internal/wire/testdata/ExampleWithMocks/foo/wire.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// arguments (Approach A). Note that the argument's type is the interface
+// type (timer), but the concrete mock type should be passed.
+func initMockedAppFromArgs(mt timer) app {
+	panic(wire.Build(appSetWithoutMocks))
+}
+
+// initMockedApp returns an app with its mocked dependencies, created
+// via providers (Approach B).
+func initMockedApp() *appWithMocks {
+	panic(wire.Build(mockAppSet))
```
Comment (zombiezen) [inline wire/internal/wire/testdata/ExampleWithMocks/foo/wire.go]: Use the `return nil`-based forms of these to retain consistency.

File: wire/README.md (hunk 1 above)
Comment (vangent) [reply wire/README.md]: Done.

File: wire/internal/wire/testdata/ExampleWithMocks/foo/wire.go (hunk 2 above)
Comment (vangent) [reply wire/internal/wire/testdata/ExampleWithMocks/foo/wire.go]: Done.

**Pull request merged**